### ISSUE TITLE
internal/imap: assignment to entry in nil map

### ISSUE THREAD

**Issue by kortschak** [ProtonMail/proton-bridge#54]
This looks to be a problem with internal/imap/store.Message; it returns a *pmapi.Message with a nil Header.

Since this is completely repeatable (running the bridge results in a panic as soon as the client attempts to pull messages) and I have not upgraded the bridge, this is presumably something that is recently broken at the server end, though the bridge should be robust to it and not panic.

## Expected Behavior
Messages are pulled from the server.

## Current Behavior
protonmail-bridge panics

```
panic: assignment to entry in nil map

goroutine 147 [running]:
net/textproto.MIMEHeader.Set(...)
	net/textproto/header.go:22
github.com/ProtonMail/proton-bridge/pkg/message.SetBodyContentFields(0xc000774c78, 0xc00072c000)
	github.com/ProtonMail/proton-bridge@/pkg/message/header.go:98 +0x12e
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).setMessageContentType(0xc0005fc000, 0xc00072c000, 0x13cdba0, 0x1269ee0, 0x200)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_message.go:626 +0xbc
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).buildMessageInner(0xc0005fc000, 0xc00072c000, 0xc000343170, 0x0, 0x1, 0x120, 0x1101d60, 0x21a3000000000008, 0xc00074af70)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_message.go:681 +0x6d
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).buildMessage(0xc0005fc000, 0xc00072c000, 0xc00075e7b0, 0x0, 0x58, 0xc000754000, 0xb0, 0x2)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_message.go:663 +0x1b6
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).getBodyStructure(0xc0005fc000, 0x1410f40, 0xc00075e690, 0xb, 0x0, 0xc00074b9b8, 0xa77e8c)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_message.go:339 +0x12c
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).getMessage(0xc0005fc000, 0x1410f40, 0xc00075e690, 0xc0000d0fc0, 0x4, 0x6, 0x0, 0x0, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_message.go:299 +0x39c
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).ListMessages.func2(0x10339a0, 0xc000226bd0, 0x1, 0x0, 0x0, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:389 +0x33f
github.com/ProtonMail/proton-bridge/pkg/parallel.RunParallel.func2(0xc0000cae70, 0xc0000439e0, 0xc000048e40, 0xc000226cb0, 0xc000043a40)
	github.com/ProtonMail/proton-bridge@/pkg/parallel/parallel.go:79 +0xfa
created by github.com/ProtonMail/proton-bridge/pkg/parallel.RunParallel
	github.com/ProtonMail/proton-bridge@/pkg/parallel/parallel.go:76 +0x1d7
```

## Possible Solution
There is no workaround.

## Steps to Reproduce

1. Run protonmail-bridge.
2. Get client to sync
3. Bang!

## Context (Environment)
I am trying to read email.

## Detailed Description

Evolution mail client.

```
~ $ evolution --version
evolution 3.28.5-0ubuntu0.18.04.2 
~ $ protonmail-bridge --version
INFO[0000] Run app                                       appLong="Protonmail Bridge" appShort=bridge args="[protonmail-bridge --version]" build="2020-05-19T00:33:31+0200" pkg=main revision=50ed40f205 runtime=linux version=1.2.7
Protonmail Bridge version 1.2.7 (50ed40f205) 2020-05-19T00:33:31+0200
```

Note that the failure is replicated when using thunderbird.

## Possible Implementation

Have `*imapMailbox.setMessageContentType` check `m`'s validity before use?

**Comment by kortschak** [comment]
Also note from that the panic handler does not catch this since the panic is not re-panicked and there is no decoration beyond what the runtime provides.

**Comment by mikolajpp** [comment]
Seeing exact same bug. Problems started after I have switched multiple addresses mode.

**Comment by kortschak** [comment]
I made no changes, so I suspect that your change is likely to be coincidental. The only way (from a brief reading of the code) that I can see this behaviour coming is that the header is not being properly sent from the server and so it is not being unmarshaled into `Header`, leaving it `nil`. A crude workaround would be to make sure `Message()` returns a `*pmapi.Message` with a non-`nil` `Header`, but I suspect that would be papering over the issue.

**Comment by viertaxa** [comment]
Also seeing this issue.

**Comment by Eckoa** [comment]
This may be what's causing my PMBridge to close repeatedly. 
It was working fairly recently but after breaking my Linux install I reinstalled and waited a week then tried setup tonight just to have PMBridge close on initial sync. If I set it to only pull the last 2 weeks in my client it seems OK though for the moment.

**Comment by kortschak** [comment]
You should see output similar to the panic message above in your syslog if this is the cause.

**Comment by Eckoa** [comment]
> You should see output similar to the panic message above in your syslog if this is the cause.

yup seeing very similar output, panic on assignment to entry in nil map

**Comment by kortschak** [comment]
When I was replicating with thunderbird, I noticed that it did manage to pull some messages. This is consistent with your observation that if a limited set is asked for the system stays up.

**Comment by kortschak** [comment]
I have got a message from support that the fault has been reverted and confirm that bridge is working again.

This issue does indicate though that there are some issues with handling bad messages from the server that probably need to be ironed out.

**Comment by jameshoulahan** [comment]
The root cause of the issue was an API change that caused some message headers to be unparseable by the go standard library's `mail.ReadMessage` function. This change has been rolled back; investigation is underway as to why the headers became unparseable.

Bridge was not without fault -- we did not handle failure to parse headers properly, leaving the header map `nil`. This has since been fixed.
